pkg/controllers/thanos-querier/components.go

```diff
@@ -0,0 +1,170 @@
+		ensureLabels(newServiceMonitor(name, thanos.Namespace), resourceLabels),
+	}
+}
+
+func newThanosQuerierDeployment(name string, spec *msoapi.ThanosQuerier, sidecarUrls []string) *appsv1.Deployment {
+	args := []string{
+		"query",
+		"--grpc-address=127.0.0.1:10901",
+		"--http-address=127.0.0.1:9090",
+		"--log.format=logfmt",
+		"--query.auto-downsampling",
+	}
+	for _, endpoint := range sidecarUrls {
+		// args = append(args, fmt.Sprintf("--endpoint=%s", endpoint))
+		args = append(args, fmt.Sprintf("--store=%s", endpoint))
+	}
+
+	for _, rl := range spec.Spec.ReplicaLabels {
+		args = append(args, fmt.Sprintf("--query.replica-label=%s", rl))
+	}
+
+	thanos := &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: appsv1.SchemeGroupVersion.String(),
+			Kind:       "Deployment",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: spec.Namespace,
+			Labels: map[string]string{
+				"app.kubernetes.io/instance": name,
+			},
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: func(i int32) *int32 { return &i }(1),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app.kubernetes.io/instance": name,
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      name,
+					Namespace: spec.Namespace,
+					Labels: map[string]string{
+						"app.kubernetes.io/instance": name,
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "thanos-querier",
+							Args:  args,
+							Image: "quay.io/thanos/thanos:v0.22.0",
+							Ports: []corev1.ContainerPort{
+								{
+									ContainerPort: 9090,
+									Name:          "metrics",
+								},
+							},
+							TerminationMessagePolicy: "FallbackToLogsOnError",
+						},
+					},
+					NodeSelector: map[string]string{
+						"kubernetes.io/os": "linux",
+					},
+				},
+			},
+			ProgressDeadlineSeconds: func(i int32) *int32 { return &i }(300),
+		},
+	}
+
+	return thanos
+}
+
+func newServiceAccount(name string, namespace string) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "ServiceAccount",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+}
+
+func newService(name string, namespace string) *corev1.Service {
+	return &corev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "Service",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Port: 9090,
+					Name: "http",
+				},
+			},
+			Selector: map[string]string{
+				"app.kubernetes.io/instance": name,
+			},
+			Type: "ClusterIP",
+		},
+	}
+}
+
+func newServiceMonitor(name string, namespace string) *monv1.ServiceMonitor {
```

Comment:
We can resolve this in a subsequent PR, but creating a service monitor like this could be problematic because prometheus instances can be configured to reconcile SMs with certain labels. 

In the stack controller we solve this using a dedicated scrape config for targets that should always be present: https://github.com/rhobs/monitoring-stack-operator/blob/main/pkg/controllers/monitoring-stack/components.go#L510

Reply:
hmm not sure I understand fully, but I think this should still work as the thanos-controller creates all resources via server-side-apply. So any changes to fields not owned by the controller are tolerated.

Reply:
The issue is not really related to SSA, but rather to the fact that service monitor selectors can be controlled by the user. This service monitor has the labels `app.kubernetes.io/instance` so if I user sets a selector in the stack which does not cover those labels, the monitor will be dropped: https://github.com/rhobs/monitoring-stack-operator/blob/main/pkg/apis/v1alpha1/types.go#L58

We have this situation in another place, so it's fine to handle it in a subsequent issue. I just wanted to raise awareness of the problem.